mmgen/models/architectures/biggan/generator_discriminator_deep.py

```diff
@@ -0,0 +1,635 @@
+            Defaults to dict(type='BN').
+        pretrained (str | dict, optional): Path for the pretrained model or
+            dict containing information for pretained models whose necessary
+            key is 'ckpt_path'. Besides, you can also provide 'prefix' to load
+            the generator part from the whole state dict. Defaults to None.
+        reshape_style (str, optional): The reshape style for first linear
+            layer's output feature. Set to 'torch' or 'tf'. If set to 'torch',
+            we just reshape the tensor with shape (Batch, Channel, Height,
+            Width). Otherwise, we will first reshape the tensor with shape
+            (Batch, Height, Width, Channel), then swap the dimensions to
+            (Batch, Channel, Height, Width). Default to 'torch'.
+    """
+
+    def __init__(self,
+                 output_scale,
+                 noise_size=120,
+                 num_classes=0,
+                 out_channels=3,
+                 base_channels=96,
+                 block_depth=2,
+                 input_scale=4,
+                 with_shared_embedding=True,
+                 shared_dim=128,
+                 sn_eps=1e-6,
+                 init_type='ortho',
+                 concat_noise=True,
+                 act_cfg=dict(type='ReLU', inplace=False),
+                 conv_cfg=dict(type='Conv2d'),
+                 upsample_cfg=dict(type='nearest', scale_factor=2),
+                 with_spectral_norm=True,
+                 auto_sync_bn=True,
+                 blocks_cfg=dict(type='BigGANDeepGenResBlock'),
+                 arch_cfg=None,
+                 out_norm_cfg=dict(type='BN'),
+                 pretrained=None,
+                 reshape_style='torch'):
+        super().__init__()
+        self.noise_size = noise_size
+        self.num_classes = num_classes
+        self.shared_dim = shared_dim
+        self.with_shared_embedding = with_shared_embedding
+        self.output_scale = output_scale
+        self.arch = arch_cfg if arch_cfg else self._get_default_arch_cfg(
+            self.output_scale, base_channels)
+        self.input_scale = input_scale
+        self.concat_noise = concat_noise
+        self.blocks_cfg = deepcopy(blocks_cfg)
+        self.upsample_cfg = deepcopy(upsample_cfg)
+        self.block_depth = block_depth
+        self.reshape_style = reshape_style
+
+        # Validity Check
+        # If 'num_classes' equals to zero, we shall set 'with_shared_embedding'
+        # to False.
+        if num_classes == 0:
+            assert not self.with_shared_embedding
+            assert not self.concat_noise
+        else:
+            if not self.with_shared_embedding:
+                # If not `with_shared_embedding`, we will use `nn.Embedding` to
+                # replace the original `Linear` layer in conditional BN.
+                # Meanwhile, we do not adopt split noises.
+                assert not self.concat_noise
```

Comment:
What's the relation between `concat_noise` and `split noises`?

Reply:
Will, In BigGAN-Deep, if we set `concat_noise` to True, the noise will be concatenated with class vector and then passed to the linear layer and ResBlocks. Otherwise, the noise will be passed to the linear layer alone, and the one-hot label will be passed to the ResBlocks. In BigGAN, if we set 'split_noise' to True, the noise will be split into slots and sent to the linear layer and ResBlock(concatenated with class vector) separately. Otherwise, the noise will be sent to the linear layer only. In the official PyTorch [implementation](https://github.com/ajbrock/BigGAN-PyTorch), both arguments are named with `hier`.